plugins/se/se.cpp

```diff
@@ -0,0 +1,321 @@
+		return true;
+	}
+
+	if (initialize(pids, L"left4dead.exe")) {
+		return true;
+	}
+
+	return false;
+}
+
+static int tryLock(const std::multimap<std::wstring, unsigned long long int> &pids) {
+	if (!tryInit(pids)) {
+		return false;
+	}
+
+	const auto engine = getModuleAddr(isWin32 ? L"engine.dll" : L"engine.so");
+	if (!engine) {
+		return false;
+	}
+
+	const auto client = getModuleAddr(isWin32 ? L"client.dll" : L"client.so");
+	if (!client) {
+		return false;
+	}
+
+	const auto engineInterfaces = getInterfaces(engine);
+
+	const auto engineClient = getInterfaceAddress("VEngineClient013", engineInterfaces);
+	if (!engineClient) {
+		return false;
+	}
+
+	localClient = getLocalClient(engineClient);
+	if (!localClient) {
+		return false;
+	}
+
+	signOnStateOffset = getSignOnStateOffset(engineClient);
+	if (!signOnStateOffset) {
+		return false;
+	}
+
+	const auto signOnState = peekProc<uint32_t>(localClient + signOnStateOffset);
+	if (signOnState != 6) {
+		return false;
+	}
+
+	const auto clientInterfaces = getInterfaces(client);
+
+	const auto clientEntityList = getInterfaceAddress("VClientEntityList003", clientInterfaces);
+	if (!clientEntityList) {
+		return false;
+	}
+
+	localPlayer = getLocalPlayer(localClient, clientEntityList, engineClient);
+	if (!localPlayer) {
+		return false;
+	}
+
+	// "pl" is the offset to the internal player class.
+	const auto pl = getDataVarFromEntity("pl", localPlayer);
+	if (!pl) {
+		return false;
+	}
+
+	// "deadflag" is the offset relative to "pl".
+	//
+	// For some reason it's missing from Left 4 Dead 2 on Linux, along with many other datavars we don't need.
+	// The offset (0x4) has never changed across the various engine versions released during the years (10+), thus we force it.
+	const auto deadflag = getDataVarFromEntity("deadflag", localPlayer);
+	deadFlagOffset = pl + (deadflag ? deadflag : 0x4);
+
+	originPositionOffset = getDataVarFromEntity("m_vecAbsOrigin", localPlayer);
+	if (!originPositionOffset) {
+		return false;
+	}
+
+	eyesPositionOffsetOffset = getDataVarFromEntity("m_vecViewOffset", localPlayer);
+	if (!eyesPositionOffsetOffset) {
+		return false;
+	}
+
+	rotationOffset = getDataVarFromEntity("m_angAbsRotation", localPlayer);
+	if (!rotationOffset) {
+		return false;
+	}
+
+	netInfoOffset = getNetInfoOffset(localClient, engineClient);
+	if (!netInfoOffset) {
+		return false;
+	}
+
+	levelNameOffset = getLevelNameOffset(engineClient);
+	if (!levelNameOffset) {
+		return false;
+	}
+
+	float avatarPos[3], avatarFront[3], avatarTop[3];
+	float cameraPos[3], cameraFront[3], cameraTop[3];
+	std::string context;
+	std::wstring identity;
+
+	if (fetch(avatarPos, avatarFront, avatarTop, cameraPos, cameraFront, cameraTop, context, identity)) {
+		return true;
+	} else {
+		generic_unlock();
+		return false;
+	}
+}
+
+static const std::wstring longDesc() {
+	return std::wstring(L"Supports Left 4 Dead and Left 4 Dead 2 with context and identity support.");
```

Comment:
Shouldn't the description be more general? As far as I understood, this works for all Source-engine games, doesn't it? If so, I would state exactly that in the description (potentially listing l4d and l4d2 as specific examples)

Reply:
The plugin should indeed work for all Source Engine games (maybe with a few minor changes for some versions), but the game is detected from the process' name.

Right now only `left4dead2.exe` and `left4dead.exe` are specified on Windows.

On Linux `hl2_linux` is specified as well. The name is used by many games.